ansicolor.js

```diff
@@ -198,131 +213,220 @@ const TEXT    = 0,
       BRACKET = 1,
       CODE    = 2
 
-function rawParse (s) {
-    
-    let state = TEXT, buffer = '', text = '', code = '', codes = []
-    let spans = []
-
-    for (let i = 0, n = s.length; i < n; i++) {
-
-        const c = s[i]
+class Span {
+  constructor(code, text) {
+    this.code = code;
+    this.text = text;
+
+    // Those are added in the actual parse, this is done for performance reasons to have the same hidden class
+    this.css = "";
+    this.color = "";
+    this.bgColor = "";
+    this.bold = undefined;
+    this.inverse = undefined;
+    this.italic = undefined;
+    this.underline = undefined;
+    this.bright = undefined;
+    this.dim = undefined;
+  }
+}
 
-        buffer += c
+// getString as function instead of string to allow garbage collection
+function* rawParse(getString) {
+  const stateObject = {
+    state: TEXT,
+    buffer: "",
+    text: "",
+    code: "",
+    codes: [],
+  };
 
-        switch (state) {
+  const ONE_MB = 1048576;
 
-            case TEXT:
-                if (c === '\u001b') { state = BRACKET; buffer = c; }
-                else                { text += c }
-                break
+  const chunks = splitStringToChunksOfSize(getString(), ONE_MB);
 
-            case BRACKET:
-                if (c === '[') { state = CODE; code = ''; codes = [] }
-                else           { state = TEXT; text += buffer }
-                break
+  while (chunks.length > 0) {
+    const chunk = chunks.shift();
```

Comment:
The shift here is important so we don't have any reference to the data

Reply:
just a note, but shift is O(n) and causes the contents of the arr to be moved which makes this loop On^2. Depending on the amount of chunks you are expecting, you may be able to squeeze some extra perf by using a linked list

Reply:
Not necessarily, even thought linked list are theoretically better for this they might not be, the case, I tried once to move nodejs webstream internal from using array to linked list and it was actually worst

what we can do instead is set the value to undefined so it can be garbage collected and we won't change the array size

Reply:
Correct (hence my comment about chunk count). Context is important and I don't have much of it here, I wanted to flag this mostly for awareness. 

In many cases when dealing with small iterators, micro-optimizations like these are redundant. What I have seen happen with code like this is that even though you don't see any improvement, someone in the future might run into some obscure edge case (maybe by processing something with an insanely high chunk count) and experience degraded perf, but you are the lib author so you know if this is ever likely to happen (if ever) and why you imo don't need to even handle it. 

I wanted to raise this for awareness (sorry if I sidetracked the discussion), but I think I should add that you did a very good job on the entire PR 🙏

Reply:
I'm actually not the lib author

Reply:
I think it will be okay for sane uses (like 1k chunks), as the reallocating and copying of such an array would be peanuts compared to processing of chunks themselves...

Reply:
> The shift here is important so we don't have any reference to the data once we finish processing

Would be reasonable to add a comment in the code, so it won't trigger whoever will be reading that code.

Reply:
assigned to undefined and added comment